### ISSUE TITLE
Allow build step to Authenticate via OIDC and Assume Role for Lookups during Synthesis 

### ISSUE THREAD

**Issue by awshilpa** [cdklabs/cdk-pipelines-github#174]
**Issue Description** - During synthesis and build job execution time, we require to do vpc lookup as under 
const sharedVpc = Vpc.fromLookup(this, `SharedVpc-${name}`, {
          vpcName: sharedVpcName,
        });

For that I need to do authentication into my target AWS environment via OIDC, I can very well add a pre-build steps to Authenticate via OIDC and Assume role using awscredentials step action in my cdk code, but as the permissions set at job level don't provide any id token write access, and neither I am able to extend job level structure in my cdk code so it will help if we can add this id-token write permission in the jobForBuildStep function

**Error** - [Error at /Dev/PrefillApiStack/prefill-api] Could not assume role in target account using current credentials (which are for account ****) User: arn:aws:sts::<accountid>:assumed-role/Dev-github-runners-GitHubRunnerRole****/I-******** is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::<accountidtarget>:role/***-lookup-role-**-ap-southeast-2 . Please make sure that this role exists in the account. If it doesn't exist, (re)-bootstrap the environment with the right '--trust', using the latest version of the CDK CLI.

**Analysis done so far and potential fix** - There is a need to Authenticate via OIDC and Assume role during Synthesis in order to perform cdk look ups. The current permissions at job level for Synthesis /Build is only token read and it needs to be have id-token write permission for authenticating via OIDC role. Hence added the fix/change in pipeline.ts in jobForBuildStep function to add permission idToken: this.useGitHubActionRole ? github.JobPermission.WRITE : github.JobPermission.NONE

Once this is done, I added a pre build step to my synth as under and it fixed the issue.

```ts
private stepToAssumeLookUpRole(env:Environment): github.JobStep[] {
    return [awsCredentialStep('Authenticate Via OIDC Role', {
      region: env.region ?? 'ap-southeast-2',
      gitHubActionRoleArn: this.gitHubActionRoleArnValue,
    }), awsCredentialStep('Assume CDK Lookup Role', {
      region:  env.region ?? 'ap-southeast-2',
      accessKeyId: '${{ env.AWS_ACCESS_KEY_ID }}',
      secretAccessKey: '${{ env.AWS_SECRET_ACCESS_KEY }}',
      sessionToken: '${{ env.AWS_SESSION_TOKEN }}',
      roleToAssume: 'arn:aws:iam::'+env.account+':role/***-lookup-role-'+env.account+'-ap-southeast-2',
    })
    ];
  }
```

**Comment by kaizen3031593** [comment]
Thanks for the issue, @awshilpa! 

Context lookups are one of those things that we need to support, but don't need to encourage.
For context, here's what the `@aws-cdk/pipelines` readme has to say about it: https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/pipelines/README.md#context-lookups

Basically, I'm no longer worried about an elegant solution, and I'm prepared to add your contribution to unblock you. Thanks! The other thing I'll need to add (in a separate PR) is a readme documentation for how to handle context lookups.
